walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,220 @@
+
+    if (optionalAttendees.size() == 0) {
+      return available;
+    }
+
+    LinkedList<Event> applicableEventsOnlyOfOptional = removeEventsWithNoAttendees(optionalAttendees, sortedEvents);
+    LinkedList<Event> optionalNonOverlapping = sortedEventsIntoNonOverlappingSortedEvents(optionalAttendees, applicableEventsOnlyOfOptional);
+    ArrayList<TimeRange> availableWithOptional = availableTimesFromSortedNonOverlappingEvents(optionalNonOverlapping, request);
+    
+    if (available.size() == 0) {
+      return availableWithOptional;
+    }
+    
+    ArrayList<TimeRange> finalTimeRanges = mergeTimeRangesThatIntersect(availableWithOptional, available);
+    return finalTimeRanges;
+  }
+   
+  /*
+   * Function to merge time ranges that intersect between two lists of time ranges
+   * 
+   * Runtime: O(n*logm) where 'n' is the length of optional and 'm' is the length of mandatory
+   * 
+   * @param optional    A list of time ranges in which for every one we'll search for an intersection
+   * @param mandatory   A list of time ranges to search within
+   * 
+   * return   A list of time ranges where all ranges are within both input lists
+   * 
+  */
+  private ArrayList<TimeRange> mergeTimeRangesThatIntersect(ArrayList<TimeRange> optional, ArrayList<TimeRange> mandatory) {
+    ArrayList<TimeRange> finalTimeRanges = new ArrayList<>();
+
+    ArrayList<Integer> startValues = new ArrayList<>();
+    for (TimeRange mandatoryTimeRange: mandatory) {
+      startValues.add(mandatoryTimeRange.start());
+    }
+
+    for (TimeRange optionalTimeRange: optional) {
+      int start = optionalTimeRange.start();
+      int end = optionalTimeRange.end();
+
+      // This only works because java's binary search returns (-(insertion_point) - 1) 
+      // if not found, meaning that the negative is 2 indexes beyond what the closest 
+      // lesser than element exists
+      int index = Collections.binarySearch(startValues, start);
+      index = index >= 0 ? index : (index + 2) * -1;
+  
+      if (index != -1) {
+        int mandatoryEnd = mandatory.get(index).end();
+        if (end <= mandatoryEnd) {
+          finalTimeRanges.add(optionalTimeRange);
+        }
+      }
+    }
+
+    return finalTimeRanges.size() != 0 ? finalTimeRanges : mandatory;
+  }
+
+  /*
+   * Removes events in which no attendees intersect between the event and requested attendees
+   * 
+   * Runtime O(n*q) where 'n' is the length of events and 'q' is the length of the attendees
+   * 
+   * @param attendes    A hashset of all applicable attendees
+   * @param events      A sorted list of the events to merge
+   * 
+   * return   A list of events where all events have at least one attendee from the set
+   *  
+  */
+  private LinkedList<Event> removeEventsWithNoAttendees(HashSet<String> attendees, LinkedList<Event> events) {
+    LinkedList<Event> sequentialEvents = new LinkedList<>(events);
+    ListIterator listIterator = sequentialEvents.listIterator();
```

Comment:
Despite what I said about it weird to have for-loops without increment conditions, I think in this case it makes sense just to avoid declaring `listIterator` outside of the scope of the loop.

`for (ListIterator<Event> it = sequentialEvents.listIterator(); it.hasNext();)`

Same advise goes for the other two iterator loops in this PR.